### PR TITLE
Sentry fixes

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
Sentry fixes for quota usage and iOS extensions.

I tested the Share Extension and it worked well.  The things that we need to test include: 
 * Notification Extension
 * Test a build produced by CircleCI
 * Verify if events are being collected on the MM Sentry Dashboard

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48756
https://mattermost.atlassian.net/browse/MM-48758

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/utils/sentry.ts [hunk 1]
```diff
@@ -39,9 +40,20 @@ export function initializeSentry() {
         return;
     }
 
+    const isBetaApp = DeviceInfo.getBundleId && DeviceInfo.getBundleId().includes('rnbeta');
+    const mmConfig = {
+        environment: isBetaApp ? 'beta' : 'production',
+        tracesSampleRate: isBetaApp ? 1.0 : 0.2,
+        sampleRate: isBetaApp ? 1.0 : 0.2,
+        attachStacktrace: isBetaApp, // For Beta, stack traces are automatically attached to all messages logged
+        logLevel: isBetaApp ? Sentry.LogLevel.Debug : Sentry.LogLevel.Error,
```
Comment (enahum) [inline app/utils/sentry.ts:49]: I still think you need soke logic in captureJsException, right?

**Comment by enahum** [comment]
@avinashlng1080 what is needed for this, let's make this happen

**Comment by avinashlng1080** [comment]
> @avinashlng1080 what is needed for this, let's make this happen

File: app/utils/sentry.ts (hunk 1 above)
Comment (avinashlng1080) [reply app/utils/sentry.ts:48]: added

File: app/utils/general/index.ts [hunk 2]
```diff
@@ -51,3 +52,5 @@ export const sortByNewest = (a: SortByCreatAt, b: SortByCreatAt) => {
 
     return 1;
 };
+
+export const isBetaApp = DeviceInfo.getBundleId && DeviceInfo.getBundleId().includes('rnbeta');
```
Comment (enahum) [inline app/utils/general/index.ts:56]: why the check for `DeviceInfo.getBundleId`?

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1208,6 +1225,8 @@
 				);
 				PRODUCT_BUNDLE_IDENTIFIER = com.mattermost.rnbeta.NotificationService;
 				PRODUCT_NAME = "$(TARGET_NAME)";
+				SENTRY_DSN = "https://8166acfeeced43de98bc11f217d871c5@o1347733.ingest.sentry.io/6734065";
+				SENTRY_ENABLED = true;
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1229]: this should not be here

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -1252,6 +1271,8 @@
 				OTHER_CFLAGS = "$(inherited)";
 				PRODUCT_BUNDLE_IDENTIFIER = com.mattermost.rnbeta.NotificationService;
 				PRODUCT_NAME = "$(TARGET_NAME)";
+				SENTRY_DSN = "https://8166acfeeced43de98bc11f217d871c5@o1347733.ingest.sentry.io/6734065";
+				SENTRY_ENABLED = true;
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1275]: this should not be here

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1301,6 +1322,8 @@
 				MTL_FAST_MATH = YES;
 				PRODUCT_BUNDLE_IDENTIFIER = com.mattermost.rnbeta.MattermostShare;
 				PRODUCT_NAME = "$(TARGET_NAME)";
+				SENTRY_DSN = "https://8166acfeeced43de98bc11f217d871c5@o1347733.ingest.sentry.io/6734065";
+				SENTRY_ENABLED = true;
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1326]: this should not be here

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -1351,6 +1374,8 @@
...
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1378]: this should not be here

File: ios/NotificationService/NotificationService.swift [hunk 7]
```diff
@@ -9,6 +9,11 @@ class NotificationService: UNNotificationServiceExtension {
 
   var retryIndex = 0
   
+  override init() {
+    super.init()
+    initSentryAppExt()
+  }
+
```
Comment (avinashlng1080) [inline ios/NotificationService/NotificationService.swift:16]: I'm not sure if this is the right place to init Sentry in this extension.  Can you please confirm?

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -1364,18 +1389,28 @@
...
 				CLANG_WARN_ENUM_CONVERSION = YES;
 				CLANG_WARN_INFINITE_RECURSION = YES;
 				CLANG_WARN_INT_CONVERSION = YES;
+				CLANG_WARN_NON_LITERAL_NULL_CONVERSION = YES;
+				CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF = YES;
+				CLANG_WARN_OBJC_LITERAL_CONVERSION = YES;
 				CLANG_WARN_OBJC_ROOT_CLASS = YES_ERROR;
+				CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER = YES;
+				CLANG_WARN_RANGE_LOOP_ANALYSIS = YES;
+				CLANG_WARN_STRICT_PROTOTYPES = YES;
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1413]: what are this changes?

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -1417,18 +1452,28 @@
...
```
Comment (enahum) [inline ios/Mattermost.xcodeproj/project.pbxproj:1476]: same question

File: ios/Sentry.swift [hunk 10]
```diff
@@ -0,0 +1,33 @@
+//
```
Comment (enahum) [inline ios/Sentry.swift:1]: file location

File: ios/Sentry.swift [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+        // Features turned off by default, but worth checking out
+        options.enableAppHangTracking = true
+        options.enableFileIOTracking = true
+        options.enableCoreDataTracking = true
+        options.enableCaptureFailedRequests = true
+      }
+    }
+  }
+}
```
Comment (enahum) [inline ios/Sentry.swift:29]: format

File: ios/Sentry.swift [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+import Sentry
+
+func initSentryAppExt(){
+if let SENTRY_ENABLED = Bundle.main.infoDictionary?["SENTRY_ENABLED"] as? String,
+   let SENTRY_DSN = Bundle.main.infoDictionary?["SENTRY_DSN_IOS"] as? String {
+    if(SENTRY_ENABLED=="true"){
+      SentrySDK.start { options in
+        options.dsn = SENTRY_DSN
+        options.debug = true // Enabled debug when first installing is always helpful
```
Comment (enahum) [inline ios/Sentry.swift:19]: why is it helpful?

File: app/utils/sentry.ts [hunk 13]
```diff
@@ -39,9 +40,19 @@ export function initializeSentry() {
         return;
     }
 
+    const mmConfig = {
+        environment: isBetaApp ? 'beta' : 'production',
+        tracesSampleRate: isBetaApp ? 1.0 : 0.2,
+        sampleRate: isBetaApp ? 1.0 : 0.2,
+        attachStacktrace: isBetaApp, // For Beta, stack traces are automatically attached to all messages logged
+        logLevel: isBetaApp ? Sentry.LogLevel.Debug : Sentry.LogLevel.Error,
```
Comment (enahum) [inline app/utils/sentry.ts:48]: is LogLevel.Error only recording fatal errors?

File: app/utils/general/index.ts (hunk 2 above)
Comment (enahum) [reply app/utils/general/index.ts:56]: This constant for `isBetaApp` can also be used in `init/push_notifications.ts#L251`

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/utils/general/index.ts (hunk 2 above)
Comment (avinashlng1080) [reply app/utils/general/index.ts:56]: corrected

File: app/utils/sentry.ts (hunk 13 above)
Comment (avinashlng1080) [reply app/utils/sentry.ts:48]: I believe so, other values would be too noisy for production

File: ios/Sentry.swift (hunk 12 above)
Comment (avinashlng1080) [reply ios/Sentry.swift:19]: removed

File: app/utils/sentry.ts (hunk 13 above)
Comment (enahum) [reply app/utils/sentry.ts:48]: I believe so is not really an answer hahahaha, I think you should check for is fatal in the capture JS exception function

File: ios/Sentry.swift (hunk 10 above)
Comment (avinashlng1080) [reply ios/Sentry.swift]: fixed

File: ios/Sentry.swift (hunk 11 above)
Comment (avinashlng1080) [reply ios/Sentry.swift]: fix indentation

File: ios/Mattermost.xcodeproj/project.pbxproj (hunk 3 above)
Comment (avinashlng1080) [reply ios/Mattermost.xcodeproj/project.pbxproj]: corrected

File: ios/Mattermost.xcodeproj/project.pbxproj (hunk 4 above)
Comment (avinashlng1080) [reply ios/Mattermost.xcodeproj/project.pbxproj]: corrected

File: ios/Mattermost.xcodeproj/project.pbxproj (hunk 5 above)
Comment (avinashlng1080) [reply ios/Mattermost.xcodeproj/project.pbxproj]: corrected

File: ios/Mattermost.xcodeproj/project.pbxproj (hunk 6 above)
Comment (avinashlng1080) [reply ios/Mattermost.xcodeproj/project.pbxproj]: corrected

File: ios/Mattermost.xcodeproj/project.pbxproj (hunk 8 above)
Comment (avinashlng1080) [reply ios/Mattermost.xcodeproj/project.pbxproj]: removed - I guess they got there when I was installing the swift packages and did some wrong manipulation on my side.  I removed them

File: app/utils/sentry.ts [hunk 14]
```diff
@@ -39,9 +40,18 @@ export function initializeSentry() {
         return;
     }
 
+    const mmConfig = {
+        environment: isBetaApp ? 'beta' : 'production',
+        tracesSampleRate: isBetaApp ? 1.0 : 0.2,
+        sampleRate: isBetaApp ? 1.0 : 0.2,
+        attachStacktrace: isBetaApp, // For Beta, stack traces are automatically attached to all messages logged
```
Comment (avinashlng1080) [inline app/utils/sentry.ts:47]: @enahum  I ended removing the previous `loglevel` that was here.  That property didn't exist and for the fatal errors, I'm checking them like that https://github.com/mattermost/mattermost-mobile/blob/d10d8cda7c6a22d4ab849836b4b33cabbd5d4782/app/utils/error_handling.ts#L47

**Comment by enahum** [comment]
@avinashlng1080 tests are failing

File: app/utils/general/index.ts (hunk 2 above)
Comment (enahum) [reply app/utils/general/index.ts:56]: I don't see the correction

File: app/utils/sentry.ts (hunk 14 above)
Comment (enahum) [reply app/utils/sentry.ts:47]: Right, but here we should also add it

**Review by enahum — CHANGES_REQUESTED** [review]
Enjoy your PTO, I'll take care of the rest

File: app/utils/error_handling.ts [hunk 15]
```diff
@@ -42,7 +43,10 @@ class JavascriptAndNativeErrorHandler {
         }
 
         logWarning('Handling Javascript error', e, isFatal);
-        captureJSException(e, isFatal);
+
+        if (isBetaApp || (!isBetaApp && isFatal)) {
```
Comment (larkox) [inline app/utils/error_handling.ts:47]:
```suggestion
        if (isBetaApp || isFatal)) {
```

`X || (!X && Y)` is the same as `X || Y`)

**Review by larkox — APPROVED** [review]
LGTM, just one nit

File: ios/NotificationService/NotificationService.swift (hunk 7 above)
Comment (enahum) [reply ios/NotificationService/NotificationService.swift:16]: it looks to be the right place.. in essence, every time a notification is received, sentry gets initialized.

**Review by enahum — APPROVED** [review]

**Comment by enahum** [comment]
Removing QA from this PR, there is no way in hell they'll be able to test it.

**Pull request merged**